### PR TITLE
Implement flux reporter

### PR DESCRIPTION
**Pull request by paroxp**
## What

We need to be able to tell what deployment is currently deploy and be
able to tell the difference if it changes.

This app will be monitoring the HelmRelease resource on Kubernetes API
from Flux CRD, and compile some useful information for us to look at.

This application needs to be running on Kubernetes. Making it into a
docker image should help in artefact distribution.

We'd like to be certain some of the dependencies won't be changing any
time soon, meaning the application may break the next build.

I am against vendor. However, it means the docker image can build faster
and we don't need to worry about some of these go amiss or misalign
with lock file.

## How to review

- Code review
- [Deploy to your cluster](https://github.com/alphagov/gsp-terraform-ignition/pull/51)
- Visit the reporter

### CODE REVIEW

File: src/main.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+			log.Panicf("unable to retreive data: %s\n", err)
+		}
+
+		var releaseList flux.HelmReleaseList
+		err = json.Unmarshal(data, &releaseList)
+		if err != nil {
+			log.Panicf("unable to unmarshal data: %s\n", err)
+		}
+
+		cache = []releaseEntry{}
```
[the remark this answers is not in the document]
Comment (paroxp) [reply src/main.go]: I'm not sure if that's really a problem. I mean you're right. It's possible to have a race condition, but then the pipeline currently is set to retry for 15 minutes every minute. The cache is being rebuilt every 5s with default setting. I feel like it's not that necessary.

File: src/main.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+func healthcheckHandler(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte("OK"))
+}
+
+func helmReleasesHandler(w http.ResponseWriter, r *http.Request) {
+	data, err := json.Marshal(cache)
+	if err != nil {
+		log.Fatalf("failed to marshal json: %s\n", err)
```
[the remark this answers is not in the document]
Comment (paroxp) [reply src/main.go]: Purely based on copy paste and human typing. No real decision process involved. Do we care?

File: src/main.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var (
+	port     = kingpin.Flag("port", "Port the application should listen on.").Default("8080").OverrideDefaultFromEnvar("PORT").Short('p').String()
+	pullFreq = kingpin.Flag("pull-frequancy", "Duration of time of how often would the data be updated.").Default("5s").OverrideDefaultFromEnvar("PULL_FREQUENCY").Short('T').Duration()
+
+	cache []releaseEntry
```
[the remark this answers is not in the document]
Comment (paroxp) [reply src/main.go]: <3

File: src/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisfarms) [reply src/main.go]: yeah these failures should probably return the relevent status codes for the failed requests (ie `500` with an error message)

File: src/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (chrisfarms) [reply src/main.go]: add a mu.RWLock() around this ...and a mu.RLock() around the read

**Pull request merged**